content/blog/single-message-transforms-swiss-army-knife-of-kafka-connect.asciidoc

```diff
@@ -0,0 +1,235 @@
+You might be better off implementing a bespoke SMT which is focussing on one particular task, leveraging the full capabilities of the Java programming language.
+
+[NOTE]
+.SMT Testing
+====
+Whether you use ready-made SMTs by means of configuration, or you implement custom SMTs in Java,
+testing your work is essential.
+
+While unit tests are a viable option for basic testing of bespoke SMT implementations,
+integration tests running against Kafka Connect connectors are recommended for testing SMT configurations.
+That way you'll be sure that the SMT can process actual messages and it has been configured the way you intended to.
+
+Testcontainers and the Debezium https://debezium.io/documentation/reference/integrations/testcontainers.html[support for Testcontainers] are a great foundation for setting up all the required components such as Apache Kafka, Kafka Connect, connectors and the SMTs to test.
+====
+
+A specific feature I wished for every now and then is the ability to apply SMTs only to a specific sub-set of the topics created or consumed by a connector.
+In particular if connectors create different kinds of topics (like an actual data topic and another one with with metadata),
+it can be desirable to apply SMTs only to the topics of one group but not the other.
+This requirement is captured in https://cwiki.apache.org/confluence/display/KAFKA/KIP-585%3A+Filter+and+Conditional+SMTs[KIP-585] ("Filter and Conditional SMTs"),
+please join the discussion on that one if you got requirements or feedback related to that.
+
+== Learning More
+
+There are several great presentations and blog posts out which describe in depth what SMTs,
+how you can implement your own one,
+how they are configured etc.
+
+Here are a few resources I found particularly helpful:
+
+* https://cwiki.apache.org/confluence/display/KAFKA/KIP-66%3A+Single+Message+Transforms+for+Kafka+Connect[KIP-66]: The original KIP (Kafka Improvement Proposal) that introduced SMTs
+* https://www.slideshare.net/ConfluentInc/kafka-summit-nyc-2017-singe-message-transforms-are-not-the-transformations-youre-looking-for[Singe Message Transforms are not the Transformations You're Looking For]: A great overview on SMTs, their capabilities as well as limitations, by Ewen Cheslack-Postava
+* https://medium.com/credimi-tech-blog/an-on-the-field-experience-with-kafka-connect-smts-45b170cbb5fa[A hands-on experience with Kafka Connect SMTs]: In-depth blog post on SMT use cases, things to be aware of and more, by Gian D'Uia
+
+Now, considering this wide range of use cases for SMTs, would MacGyver like and use them for implementing various tasks around Kafka Connect?
+I would certainly think so.
+But as always, the right tool for the job must be chosen: sometimes an SMT may be a great fit, another time a more flexible (and complex) stream processing solution might be preferable.
+
+Just as MacGyver, you got to make a call when to use your Swiss Army knife, duct tape or a shoe string.
```

Comment:
what happened to the "paper clip" 🙈😜

Reply:
Absolutely! In fact, I even meant to double check on this one, but then forgot.

Reply:
A pity btw. that there's no army knife emoji ;)